### PR TITLE
Set default namespace

### LINKED ISSUE #711: Apply should add the default namespace if missing
## Description

When kubectl is used to apply resources, the YAML provided doesn't need to have a namespace set, in which case it receives the `default` namespace where applicable.
However, Headlamp doesn't make set any missing namespace, and as such, applying such resources will fail.

## Impact

Users may be confused about why they've copied an example resource code from the K8s docs (which have no namespace set) and it doesn't work.

## Expected behavior

Headlamp should set up the default namespace for any resources needing it.

### PR DESCRIPTION
**Pull request by joaquimrocha**
fixes #711 .

How to test:
- [ ] Choose a namespaced resource config, like the [ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/#configmaps-and-pods) without setting a namespace in its definition, and apply it from the + button in the UI: it should succeed

### CODE REVIEW

**Review by illume — COMMENTED** [review]
With kubectl there is a default namespace, but also a namespace context. So someone copy/pasting might also have their context set to something else than the default. https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#setting-the-namespace-preference

I wonder if we can make the namespace that will be applied more obvious in the UI?

Will people be able to apply under a different namespace? So if they're browsing in a different namespace perhaps, then apply? What happens then if they remove the namespace and apply? In this case, I expect a mistake will be made.

Not all objects are in a namespace.  https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#not-all-objects-are-in-a-namespace In this case, adding a default namespace will break things when trying to apply to things not in any namespace.

File: frontend/src/lib/k8s/apiProxy.ts [hunk 1]
```diff
@@ -743,6 +744,18 @@ function combinePath(base: string, path: string) {
 }
 
 export async function apply(body: KubeObjectInterface): Promise<JSON> {
+  // Check if the default namespace is needed. And we need to do this before
+  // getting the apiEndpoint because it will affect the endpoint itself.
+  const { namespace } = body.metadata;
+  if (!namespace) {
```
[the remark this answers is not in the document]
Comment (joaquimrocha) [reply frontend/src/lib/k8s/apiProxy.ts:750]: I think it's only for apply in the sense that if you modify an object, it should already have a namespace (if it applies).

File: frontend/src/lib/k8s/apiProxy.ts [hunk 2]
```diff
@@ -743,6 +744,18 @@ function combinePath(base: string, path: string) {
...
+    const knownResource = ResourceClasses[body.kind];
+    if (knownResource?.isNamespaced) {
+      // Clone the param to avoid modifying the original object.
+      body = _.cloneDeep(body);
```
[the remark this answers is not in the document]
Comment (joaquimrocha) [reply frontend/src/lib/k8s/apiProxy.ts:754]: I'll sort this one out.

**Comment by joaquimrocha** [comment]
@illume ,

> With kubectl there is a default namespace, but also a namespace context. So someone copy/pasting might also have their context set to something else than the default. https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#setting-the-namespace-preference
> 
> I wonder if we can make the namespace that will be applied more obvious in the UI?

We use the namespace as a filter in the UI only, though it persists when changing views.
Maybe we should indeed make it a more evidently global setting. Quick thought: I think we'll change/move the cluster chooser button when we have the global overview, so in its place we could have a namespace selector.

> Will people be able to apply under a different namespace? So if they're browsing in a different namespace perhaps, then apply? What happens then if they remove the namespace and apply? In this case, I expect a mistake will be made.

I think that right now it's not evident that a namespace is selected, since it's only in table views where it's set up.
So I don't expect users to select a namespace in a table view, and then click the + icon in a different area but expect the namespace to stick.

Maybe we can make this evident in the edit dialog, but this PR is about avoiding failure.

> Not all objects are in a namespace. https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#not-all-objects-are-in-a-namespace In this case, adding a default namespace will break things when trying to apply to things not in any namespace. It is possible to query which resources aren't in a namespace, so maybe we can use that (see the URL for how to do this query).

Right. My patch only adds the namespace if we know that the resource we're applying needs one (if we don't know the resource, or the resource has no namespace, like ClusterRole, then it will not set one).

**Pull request merged**